services/galley/src/Galley/Intra/Util.hs

```diff
@@ -16,56 +18,88 @@
+import Control.Lens (view, (^.))
+import Control.Monad.Catch
 import Control.Retry
 import qualified Data.ByteString.Lazy as LB
 import Data.Misc (portNumber)
 import Data.Text.Encoding (encodeUtf8)
 import qualified Data.Text.Lazy as LT
-import Galley.App
 import Galley.Effects
+import Galley.Env
 import Galley.Options
 import Imports
+import Network.HTTP.Types
+import Polysemy
+import qualified Polysemy.Reader as P
 import Util.Options
 
-brigReq :: Galley r (ByteString, Word16)
-brigReq = do
-  h <- encodeUtf8 <$> view (options . optBrig . epHost)
-  p <- portNumber . fromIntegral <$> view (options . optBrig . epPort)
-  return (h, p)
+data IntraComponent = Brig | Spar | Gundeck
+  deriving (Show)
 
-sparReq :: Galley r (ByteString, Word16)
-sparReq = do
-  h <- encodeUtf8 <$> view (options . optSpar . epHost)
-  p <- portNumber . fromIntegral <$> view (options . optSpar . epPort)
-  return (h, p)
+componentName :: IntraComponent -> String
+componentName Brig = "brig"
+componentName Spar = "spar"
+componentName Gundeck = "gundeck"
 
--- gundeckReq lives in Galley.Intra.Push
+componentRequest :: IntraComponent -> Opts -> Request -> Request
+componentRequest Brig o =
+  host (encodeUtf8 (o ^. optBrig . epHost))
+    . port (portNumber (fromIntegral (o ^. optBrig . epPort)))
+componentRequest Spar o =
+  host (encodeUtf8 (o ^. optSpar . epHost))
+    . port (portNumber (fromIntegral (o ^. optSpar . epPort)))
+componentRequest Gundeck o =
+  host (encodeUtf8 $ o ^. optGundeck . epHost)
+    . port (portNumber $ fromIntegral (o ^. optGundeck . epPort))
+    . method POST
+    . path "/i/push/v2"
+    . expect2xx
 
-call0 :: LT.Text -> (Request -> Request) -> Galley0 (Response (Maybe LB.ByteString))
-call0 n r = liftGalley0 $ recovering x1 rpcHandlers (const (rpc n r))
+embedIntra ::
+  Members '[Embed IO, P.Reader Env] r =>
+  IntraM a ->
+  Sem r a
+embedIntra action = do
+  env <- P.ask
+  embed $ runHttpT (env ^. manager) (runReaderT (unIntraM action) env)
 
-callBrig :: Member BrigAccess r => (Request -> Request) -> Galley r (Response (Maybe LB.ByteString))
-callBrig r = liftGalley0 $ call0 "brig" r
+newtype IntraM a = IntraM {unIntraM :: ReaderT Env Http a}
+  deriving
+    ( Functor,
+      Applicative,
+      Monad,
+      MonadIO,
+      MonadHttp,
+      MonadThrow,
+      MonadCatch,
+      MonadMask,
+      MonadReader Env,
+      MonadUnliftIO
+    )
 
-callSpar :: Member SparAccess r => (Request -> Request) -> Galley r (Response (Maybe LB.ByteString))
-callSpar r = liftGalley0 $ call0 "spar" r
+instance HasRequestId IntraM where
+  getRequestId = IntraM $ view reqId
 
-callBot :: Member BotAccess r => (Request -> Request) -> Galley r (Response (Maybe LB.ByteString))
-callBot r = liftGalley0 $ call0 "brig" r
+call ::
+  IntraComponent ->
+  (Request -> Request) ->
+  IntraM (Response (Maybe LB.ByteString))
+call comp r = do
+  o <- view options
+  let r0 = componentRequest comp o
+  let n = LT.pack (componentName comp)
```

Comment:
Why? It doesn't make any difference. `LT.pack` is still going to be called under the hood.

Reply:
This is nothing important; I was just thinking why not put `LT.pack` in the function definition instead of calling it at every use site of `componentName`.

Reply:
But there is only one use site, isn't there?